src/iop/prover.rs

```diff
@@ -45,14 +48,26 @@ impl<'a, F: PrimeField> ProverState<'a, F> {
     pub fn public_input(&self) -> Vec<F> {
         IOP::unformat_public_input(&self.formatted_input_assignment)
     }
-    /// Return the concatenation of public input (padded with zeros to match the
-    /// size of domain X) and witness variables.
-    pub fn padded_variables(&self) -> Vec<F> {
-        let padding_size = self.domain_x.size() - self.formatted_input_assignment.len();
-        let mut padded_public_input = self.formatted_input_assignment.clone();
-        padded_public_input.extend(vec![F::zero(); padding_size]);
-        let mut witness_assignment = self.witness_assignment.clone();
-        padded_public_input.append(&mut witness_assignment);
+    /// Return the variable vector, with input variables and witness variables already indexed
+    /// according to the treatment of the input domain `domain_x` as a subdomain of the full
+    /// Lagrange domain `domain_h`.
+    pub fn full_variable_vector(&self) -> Vec<F> {
+        let domain_x_size = self.domain_x.size();
+        let mut padded_public_input = vec![F::zero(); self.domain_h.size()];
```

Comment:
The reason is that slices are just a pointer (plus length) to a contiguous sequence of data. A function cannot return references to data allocated inside the function itself, because they would be dangling references after the function returns.